File: packages/core/src/kilter.ts
import type { Ascent, BoardAuth, ConnectOptions, ConnectResult } from "./types.js";
import { BoardError } from "./types.js";
import { DIFFICULTY_GRADES, gradeForDifficulty, type DifficultyGrade } from "./difficulty.js";
import { addSetCookies, jarToHeader, type CookieJar } from "./http.js";

// The Kilter app's backend (kiltergrips.com) since Kilter left Aurora in 2025. See
// docs/kilter-new-api.md. Auth is Keycloak (OIDC + PKCE); the logbook comes from the portal REST API
// (GET /api/logs), enriched by the server with climb names and grades.

export const KILTER_HOSTS = {
  idp: "https://idp.kiltergrips.com",
  api: "https://portal.kiltergrips.com",
  sync: "https://sync1.kiltergrips.com",
} as const;

const REALM = "kilter";
const CLIENT_ID = "kilter";
const REDIRECT_URI = "com.kiltergrips:/oauthredirect";
const APP_UA = "Dart/3.10 (dart:io)";

export interface KilterTokens {
  accessToken: string;
  refreshToken: string;
  expiresIn?: number;
}

function base64url(bytes: Uint8Array): string {
  let s = "";
  for (const b of bytes) s += String.fromCharCode(b);
  return btoa(s).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function randomToken(len = 32): string {
  const b = new Uint8Array(len);
  crypto.getRandomValues(b);
  return base64url(b);
}

async function pkceChallenge(verifier: string): Promise<string> {
  const digest = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(verifier));
  return base64url(new Uint8Array(digest));
}

function form(data: Record<string, string>): string {
  return Object.entries(data)
    .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(v)}`)
    .join("&");
}

function decodeEntities(s: string): string {
  return s.replace(/&amp;/g, "&").replace(/&#x2F;/g, "/").replace(/&quot;/g, '"').replace(/&#39;/g, "'");
}

// Runs the Authorization Code + PKCE flow. The password is used once and never returned or stored.
export async function kilterLogin(
  username: string,
  password: string,
  opts: ConnectOptions = {},
): Promise<KilterTokens> {
  const doFetch = opts.fetch ?? fetch;
  if (!username || !password) {
    throw new BoardError("missing-credentials", "username and password required", "kilter");
  }

  const verifier = randomToken(48);
  const authUrl =
    `${KILTER_HOSTS.idp}/realms/${REALM}/protocol/openid-connect/auth?` +
    form({
      response_type: "code",
      client_id: CLIENT_ID,
      redirect_uri: REDIRECT_URI,
      scope: "openid offline_access",
      code_challenge: await pkceChallenge(verifier),
      code_challenge_method: "S256",
      state: randomToken(16),
      nonce: randomToken(16),
    });

  let page: Response;
  try {
    page = await doFetch(authUrl, { headers: { "User-Agent": APP_UA }, redirect: "manual" });
  } catch {
    throw new BoardError("unreachable", "could not reach Kilter login", "kilter");
  }
  const jar: CookieJar = new Map();
  addSetCookies(jar, page);
  const html = await page.text();
  const action = html.match(/action=["']([^"']+login-actions\/authenticate[^"']*)["']/i)?.[1];
  if (!action) {
    throw new BoardError("unexpected-response", "Kilter login form changed (no action URL)", "kilter");
  }

  let res: Response;
  try {
    res = await doFetch(decodeEntities(action), {
      method: "POST",
      redirect: "manual",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        "User-Agent": APP_UA,
        Cookie: jarToHeader(jar),
      },
      body: form({ username, password }),
    });
  } catch {
    throw new BoardError("unreachable", "could not reach Kilter login", "kilter");
  }

  // Success is a 302 back to the app's redirect URI; a re-rendered form (200) means bad credentials.
  const location = res.headers.get("location") ?? "";
  if (res.status < 300 || res.status >= 400 || !location.startsWith(REDIRECT_URI)) {
    throw new BoardError("bad-credentials", "Incorrect Kilter email or password.", "kilter");
  }
  const code = new URL(location.replace(REDIRECT_URI, "https://x")).searchParams.get("code");
  if (!code) throw new BoardError("unexpected-response", "no authorization code returned", "kilter");

  return tokenRequest(doFetch, {
    grant_type: "authorization_code",
    code,
    code_verifier: verifier,
    redirect_uri: REDIRECT_URI,
    client_id: CLIENT_ID,
  });
}

// Trades a stored refresh token for a fresh access token, so re-syncs need no password.
export async function kilterRefresh(refreshToken: string, opts: ConnectOptions = {}): Promise<KilterTokens> {
  return tokenRequest(opts.fetch ?? fetch, {
    grant_type: "refresh_token",
    refresh_token: refreshToken,
    client_id: CLIENT_ID,
  });
}

async function tokenRequest(doFetch: typeof fetch, body: Record<string, string>): Promise<KilterTokens> {
  let res: Response;
  try {
    res = await doFetch(`${KILTER_HOSTS.idp}/realms/${REALM}/protocol/openid-connect/token`, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded", "User-Agent": APP_UA },
      body: form(body),
    });
  } catch {
    throw new BoardError("unreachable", "could not reach Kilter token endpoint", "kilter");
  }
  if (res.status === 400 || res.status === 401) {
    const refreshing = body.grant_type === "refresh_token";
    throw new BoardError(
      refreshing ? "session-expired" : "bad-credentials",
      refreshing ? "Kilter session expired" : "Kilter login rejected",
      "kilter",
    );
  }
  if (!res.ok) throw new BoardError("unexpected-response", `token request failed (${res.status})`, "kilter");
  const json = (await res.json().catch(() => ({}))) as {
    access_token?: string;
    refresh_token?: string;
    expires_in?: number;
  };
  if (!json.access_token || !json.refresh_token) {
    throw new BoardError("unexpected-response", "token response missing tokens", "kilter");
  }
  return { accessToken: json.access_token, refreshToken: json.refresh_token, expiresIn: json.expires_in };
}

// Kilter's currentDifficultyId indexes the shared Aurora difficulty table (see difficulty.ts).
export type KilterGrade = DifficultyGrade;
export const KILTER_DIFFICULTY_GRADES = DIFFICULTY_GRADES;

// Resolves a difficulty_grade_id to its grade on every scale, or undefined if unknown.
export const kilterGrade = gradeForDifficulty;

// One entry from GET /api/logs, the authenticated user's logbook. The server joins the climb name
// and its current consensus difficulty in, so no catalog lookup is needed. Extra fields are kept.
export interface KilterLog {
  logUuid: string;
  climbUuid: string;
  climbName: string;
  angle: number;
  attempts: number;
  flashed: boolean;
  topped: boolean;
  createdAt: string;
  currentDifficultyId?: number;
  /** The user's own grade suggestion for this climb, when they submitted one. */
  climbRating?: { difficultyGradeId?: number } & Record<string, unknown>;
  [key: string]: unknown;
}

// Fetches the authenticated user's logbook. The bearer token identifies the user; there is no id in
// the path (the /api/logs/{id} form is for viewing another user's public logs).
export async function kilterFetchLogbook(accessToken: string, opts: ConnectOptions = {}): Promise<KilterLog[]> {
  const doFetch = opts.fetch ?? fetch;
  let res: Response;
  try {
    res = await doFetch(`${KILTER_HOSTS.api}/api/logs`, {
      headers: { Authorization: `Bearer ${accessToken}`, "User-Agent": APP_UA, Accept: "application/json" },
    });
  } catch {
    throw new BoardError("unreachable", "could not reach Kilter logbook", "kilter");
  }
  if (res.status === 401) throw new BoardError("session-expired", "Kilter session expired", "kilter");
  if (!res.ok) throw new BoardError("unexpected-response", `logbook request failed (${res.status})`, "kilter");
  const body = await res.json().catch(() => null);
  return Array.isArray(body) ? (body as KilterLog[]) : [];
}

// Maps one logbook entry to a normalized ascent. The consensus grade fills {@link Ascent.grade};
// the user's own suggestion, when present and different, fills {@link Ascent.userGrade}.
export function kilterLogToAscent(log: KilterLog): Ascent {
  const consensus = kilterGrade(log.currentDifficultyId);
  const ownId = log.climbRating?.difficultyGradeId;
  const own = ownId != null ? KILTER_DIFFICULTY_GRADES[ownId] : undefined;
  return {
    board: "kilter",
    climbName: log.climbName ?? "",
    date: log.createdAt,
    grade: consensus?.label,
    userGrade: own && own.label !== consensus?.label ? own.label : undefined,
    vGrade: consensus?.vGrade,
    tries: typeof log.attempts === "number" ? log.attempts : log.flashed ? 1 : undefined,
    angle: typeof log.angle === "number" ? log.angle : undefined,
    raw: log,
  };
}

// Pass credentials or a stored refresh token. Returns the (rotated) refresh token to store and the
// normalized logbook. Only topped ascents are included; raw attempt logs are skipped.
export async function connectKilter(auth: BoardAuth, opts: ConnectOptions = {}): Promise<ConnectResult> {
  const tokens =
    "token" in auth ? await kilterRefresh(auth.token, opts) : await kilterLogin(auth.username, auth.password, opts);
  const logs = await kilterFetchLogbook(tokens.accessToken, opts);
  const ascents = logs.filter((l) => l.topped !== false).map(kilterLogToAscent);
  return { board: "kilter", token: tokens.refreshToken, ascents };
}

// --- Advanced: raw board data over PowerSync ---------------------------------------------------
// The logbook above needs only the REST API. The app also streams the full board dataset (holds,
// walls, gyms, difficulty grades, hold geometry) over PowerSync; kilterPowersyncPull exposes that
// raw, for callers that want to render climbs or inspect the catalog. It is not needed for ascents.

interface SyncData {
  bucket: string;
  data: { op_id: string; op: string; object_type?: string; object_id?: string; data?: string | object }[];
}

// PowerSync rows grouped by table name (object_type), keyed by object id.
export type KilterTables = Record<string, Record<string, Record<string, unknown>>>;

// Requests a full sync (after "0") and collects rows until the initial checkpoint completes. The
// connection would otherwise stay open for live updates, so we stop once the first checkpoint lands.
export async function kilterPowersyncPull(accessToken: string, opts: ConnectOptions = {}): Promise<KilterTables> {
  const doFetch = opts.fetch ?? fetch;
  const body = {
    buckets: [{ name: "global[]", after: "0" }],
    include_checksum: true,
    raw_data: true,
    binary_data: false,
    client_id: randomUuid(),
    parameters: {},
    streams: { include_defaults: true, subscriptions: [] },
    app_metadata: {},
  };

  let res: Response;
  try {
    res = await doFetch(`${KILTER_HOSTS.sync}/sync/stream`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Accept: "application/x-ndjson",
        Authorization: `Bearer ${accessToken}`,
        "User-Agent": "powersync-dart-core/1.7.0",
      },
      body: JSON.stringify(body),
    });
  } catch {
    throw new BoardError("unreachable", "could not reach Kilter sync", "kilter");
  }
  if (res.status === 401) throw new BoardError("session-expired", "Kilter sync unauthorized", "kilter");
  if (!res.ok || !res.body) {
    throw new BoardError("unexpected-response", `sync failed (${res.status})`, "kilter");
  }

  const tables: KilterTables = {};
  for await (const line of ndjsonLines(res.body)) {
    let msg: Record<string, unknown>;
    try {
      msg = JSON.parse(line) as Record<string, unknown>;
    } catch {
      continue;
    }
    if (msg.data) applyBucketData(tables, msg.data as SyncData);
    if ("checkpoint_complete" in msg) break;
  }
  return tables;
}

function applyBucketData(tables: KilterTables, sd: SyncData): void {
  for (const op of sd.data ?? []) {
    const type = op.object_type;
    const id = op.object_id;
    if (!type || !id) continue;
    const table = (tables[type] ??= {});
    if (op.op === "REMOVE") {
      delete table[id];
      continue;
    }
    // With raw_data set, a row's payload arrives as a JSON string.
    if (typeof op.data === "string") {
      try {
        table[id] = JSON.parse(op.data) as Record<string, unknown>;
      } catch {
        table[id] = { _raw: op.data };
      }
    } else if (op.data && typeof op.data === "object") {
      table[id] = op.data as Record<string, unknown>;
    } else {
      table[id] = {};
    }
  }
}

async function* ndjsonLines(stream: ReadableStream<Uint8Array>): AsyncGenerator<string> {
  const decoder = new TextDecoder();
  let buf = "";
  // @ts-expect-error web ReadableStream is async-iterable on Node 18+ and in browsers
  for await (const chunk of stream) {
    buf += decoder.decode(chunk as Uint8Array, { stream: true });
    let nl: number;
    while ((nl = buf.indexOf("\n")) >= 0) {
      const line = buf.slice(0, nl).trim();
      buf = buf.slice(nl + 1);
      if (line) yield line;
    }
  }
  if (buf.trim()) yield buf.trim();
}

function randomUuid(): string {
  if (typeof crypto.randomUUID === "function") return crypto.randomUUID();
  const b = new Uint8Array(16);
  crypto.getRandomValues(b);
  b[6] = (b[6]! & 0x0f) | 0x40;
  b[8] = (b[8]! & 0x3f) | 0x80;
  const h = [...b].map((x) => x.toString(16).padStart(2, "0"));
  return `${h.slice(0, 4).join("")}-${h.slice(4, 6).join("")}-${h.slice(6, 8).join("")}-${h.slice(8, 10).join("")}-${h.slice(10, 16).join("")}`;
}

File: python/src/boardlink/kilter.py
from __future__ import annotations

import base64
import hashlib
import json
import re
import secrets
import uuid
from typing import Optional
from urllib.parse import parse_qs, urlparse

import requests

from .grades import parse_v_grade
from .types import Ascent, BoardError, ConnectResult

# The Kilter app's backend since it left Aurora in 2025. See docs/kilter-new-api.md. Auth is Keycloak
# (OIDC + PKCE); the logbook comes from the portal REST API (GET /api/logs), enriched by the server
# with climb names and grades.
IDP = "https://idp.kiltergrips.com"
API = "https://portal.kiltergrips.com"
SYNC = "https://sync1.kiltergrips.com"
_REALM = "kilter"
_CLIENT = "kilter"
_REDIRECT = "com.kiltergrips:/oauthredirect"
_UA = "Dart/3.10 (dart:io)"

# Kilter's static difficulty_grades reference table (ids 1-39), compact source rows of
# (id, "font/V", french, yds). Bundled rather than fetched: it changes about as often as the grade
# scales themselves. Font and V-scale are split from the compound label below.
_KILTER_GRADE_ROWS = [
    (1, "1A/V0", "2b", "5.1"), (2, "1B/V0", "2c", "5.2"), (3, "1C/V0", "3a", "5.3"),
    (4, "2A/V0", "3b", "5.3"), (5, "2B/V0", "3c", "5.4"), (6, "2C/V0", "4a", "5.5"),
    (7, "3A/V0", "4b", "5.6"), (8, "3B/V0", "4c", "5.7"), (9, "3C/V0", "5a", "5.8"),
    (10, "4A/V0", "5b", "5.9"), (11, "4B/V0", "5c", "5.10a"), (12, "4C/V0", "6a", "5.10b"),
    (13, "5A/V1", "6a+", "5.10c"), (14, "5B/V1", "6b", "5.10d"), (15, "5C/V2", "6b+", "5.11a"),
    (16, "6A/V3", "6c", "5.11b"), (17, "6A+/V3", "6c+", "5.11c"), (18, "6B/V4", "7a", "5.11d"),
    (19, "6B+/V4", "7a+", "5.12a"), (20, "6C/V5", "7b", "5.12b"), (21, "6C+/V5", "7b+", "5.12c"),
    (22, "7A/V6", "7c", "5.12d"), (23, "7A+/V7", "7c+", "5.13a"), (24, "7B/V8", "8a", "5.13b"),
    (25, "7B+/V8", "8a+", "5.13c"), (26, "7C/V9", "8b", "5.13d"), (27, "7C+/V10", "8b+", "5.14a"),
    (28, "8A/V11", "8c", "5.14b"), (29, "8A+/V12", "8c+", "5.14c"), (30, "8B/V13", "9a", "5.14d"),
    (31, "8B+/V14", "9a+", "5.15a"), (32, "8C/V15", "9b", "5.15b"), (33, "8C+/V16", "9b+", "5.15c"),
    (34, "9A/V17", "9c", "5.15d"), (35, "9A+/V18", "9c+", "5.16a"), (36, "9B/V19", "10a", "5.16b"),
    (37, "9B+/V20", "10a+", "5.16c"), (38, "9C/V21", "10b", "5.16d"), (39, "9C+/V22", "10b+", "5.17a"),
]


def _build_difficulty_grades() -> dict:
    out = {}
    for gid, label, french, yds in _KILTER_GRADE_ROWS:
        font, v_scale = label.split("/")
        out[gid] = {
            "label": label,
            "font": font,
            "v_scale": v_scale,
            "v_grade": parse_v_grade(v_scale),
            "french": french,
            "yds": yds,
        }
    return out


# difficulty_grade_id -> the grade on every scale (label, font, v_scale, v_grade, french, yds), so
# callers can pick whichever grading system they want.
KILTER_DIFFICULTY_GRADES = _build_difficulty_grades()


def _b64url(b: bytes) -> str:
    return base64.urlsafe_b64encode(b).rstrip(b"=").decode()


def _decode_entities(s: str) -> str:
    return s.replace("&amp;", "&").replace("&#x2F;", "/").replace("&quot;", '"').replace("&#39;", "'")


def kilter_login(username: str, password: str) -> tuple:
    if not username or not password:
        raise BoardError("missing-credentials", "username and password required", "kilter")
    session = requests.Session()
    verifier = _b64url(secrets.token_bytes(48))
    challenge = _b64url(hashlib.sha256(verifier.encode()).digest())
    params = {
        "response_type": "code",
        "client_id": _CLIENT,
        "redirect_uri": _REDIRECT,
        "scope": "openid offline_access",
        "code_challenge": challenge,
        "code_challenge_method": "S256",
        "state": _b64url(secrets.token_bytes(16)),
        "nonce": _b64url(secrets.token_bytes(16)),
    }

    try:
        page = session.get(
            f"{IDP}/realms/{_REALM}/protocol/openid-connect/auth",
            params=params,
            headers={"User-Agent": _UA},
            allow_redirects=False,
            timeout=30,
        )
    except requests.RequestException as e:
        raise BoardError("unreachable", "could not reach Kilter login", "kilter") from e

    action = re.search(r'action=["\']([^"\']+login-actions/authenticate[^"\']*)["\']', page.text, re.I)
    if not action:
        raise BoardError("unexpected-response", "Kilter login form changed (no action URL)", "kilter")

    try:
        r = session.post(
            _decode_entities(action.group(1)),
            data={"username": username, "password": password},
            headers={"User-Agent": _UA, "Content-Type": "application/x-www-form-urlencoded"},
            allow_redirects=False,
            timeout=30,
        )
    except requests.RequestException as e:
        raise BoardError("unreachable", "could not reach Kilter login", "kilter") from e

    # Success is a 302 back to the app's redirect URI; a re-rendered form means bad credentials.
    location = r.headers.get("location", "")
    if not (300 <= r.status_code < 400) or not location.startswith(_REDIRECT):
        raise BoardError("bad-credentials", "Incorrect Kilter email or password.", "kilter")
    code = parse_qs(urlparse(location.replace(_REDIRECT, "https://x")).query).get("code", [None])[0]
    if not code:
        raise BoardError("unexpected-response", "no authorization code returned", "kilter")

    return _token_request({
        "grant_type": "authorization_code",
        "code": code,
        "code_verifier": verifier,
        "redirect_uri": _REDIRECT,
        "client_id": _CLIENT,
    })


def kilter_refresh(refresh_token: str) -> tuple:
    return _token_request({"grant_type": "refresh_token", "refresh_token": refresh_token, "client_id": _CLIENT})


def _token_request(data: dict) -> tuple:
    try:
        r = requests.post(
            f"{IDP}/realms/{_REALM}/protocol/openid-connect/token",
            data=data,
            headers={"User-Agent": _UA, "Content-Type": "application/x-www-form-urlencoded"},
            timeout=30,
        )
    except requests.RequestException as e:
        raise BoardError("unreachable", "could not reach Kilter token endpoint", "kilter") from e
    if r.status_code in (400, 401):
        refreshing = data["grant_type"] == "refresh_token"
        raise BoardError(
            "session-expired" if refreshing else "bad-credentials",
            "Kilter session expired" if refreshing else "Kilter login rejected",
            "kilter",
        )
    if not r.ok:
        raise BoardError("unexpected-response", f"token request failed ({r.status_code})", "kilter")
    body = r.json()
    if not body.get("access_token") or not body.get("refresh_token"):
        raise BoardError("unexpected-response", "token response missing tokens", "kilter")
    return body["access_token"], body["refresh_token"]


def kilter_fetch_logbook(access_token: str) -> list:
    """Fetch the authenticated user's logbook (GET /api/logs).

    The bearer token identifies the user; there is no id in the path (the /api/logs/{id} form is for
    viewing another user's public logs). The server joins the climb name and current difficulty in.
    """
    try:
        r = requests.get(
            f"{API}/api/logs",
            headers={"Authorization": f"Bearer {access_token}", "User-Agent": _UA, "Accept": "application/json"},
            timeout=30,
        )
    except requests.RequestException as e:
        raise BoardError("unreachable", "could not reach Kilter logbook", "kilter") from e
    if r.status_code == 401:
        raise BoardError("session-expired", "Kilter session expired", "kilter")
    if not r.ok:
        raise BoardError("unexpected-response", f"logbook request failed ({r.status_code})", "kilter")
    try:
        body = r.json()
    except ValueError:
        return []
    return body if isinstance(body, list) else []


def kilter_grade(difficulty_id: Optional[int]) -> Optional[dict]:
    """Resolve a difficulty_grade_id to its grade on every scale, or None if unknown."""
    if difficulty_id is None:
        return None
    return KILTER_DIFFICULTY_GRADES.get(difficulty_id)


def kilter_log_to_ascent(log: dict) -> Ascent:
    """Map one GET /api/logs entry to a normalized ascent.

    The consensus grade fills ``grade``; the user's own suggestion, when present and different, fills
    ``user_grade``.
    """
    consensus = kilter_grade(log.get("currentDifficultyId"))
    grade = consensus["label"] if consensus else None
    own_id = (log.get("climbRating") or {}).get("difficultyGradeId")
    own = KILTER_DIFFICULTY_GRADES.get(own_id) if own_id is not None else None
    user_grade = own["label"] if own and own["label"] != grade else None
    attempts = log.get("attempts")
    angle = log.get("angle")
    return Ascent(
        board="kilter",
        climb_name=log.get("climbName") or "",
        date=log.get("createdAt") or "",
        grade=grade,
        user_grade=user_grade,
        v_grade=consensus["v_grade"] if consensus else None,
        tries=attempts if isinstance(attempts, int) else (1 if log.get("flashed") else None),
        angle=angle if isinstance(angle, int) else None,
        raw=log,
    )


def connect_kilter(username: Optional[str] = None, password: Optional[str] = None, *, token: Optional[str] = None) -> ConnectResult:
    """Pass credentials or a stored refresh token. Returns the (rotated) refresh token and the
    normalized logbook. Only topped ascents are included; raw attempt logs are skipped.
    """
    access_token, refresh_token = kilter_refresh(token) if token else kilter_login(username, password)
    logs = kilter_fetch_logbook(access_token)
    ascents = [kilter_log_to_ascent(log) for log in logs if log.get("topped") is not False]
    return ConnectResult("kilter", refresh_token, ascents)


# --- Advanced: raw board data over PowerSync ---------------------------------------------------
# The logbook above needs only the REST API. The app also streams the full board dataset (holds,
# walls, gyms, difficulty grades, hold geometry) over PowerSync; kilter_powersync_pull exposes that
# raw, for callers that want to render climbs or inspect the catalog. It is not needed for ascents.


def kilter_powersync_pull(access_token: str) -> dict:
    """Request a full sync (after "0") and collect rows until the initial checkpoint completes; the
    connection would otherwise stay open for live updates.
    """
    body = {
        "buckets": [{"name": "global[]", "after": "0"}],
        "include_checksum": True,
        "raw_data": True,
        "binary_data": False,
        "client_id": str(uuid.uuid4()),
        "parameters": {},
        "streams": {"include_defaults": True, "subscriptions": []},
        "app_metadata": {},
    }
    try:
        r = requests.post(
            f"{SYNC}/sync/stream",
            json=body,
            headers={"Accept": "application/x-ndjson", "Authorization": f"Bearer {access_token}", "User-Agent": "powersync-dart-core/1.7.0"},
            stream=True,
            timeout=60,
        )
    except requests.RequestException as e:
        raise BoardError("unreachable", "could not reach Kilter sync", "kilter") from e
    if r.status_code == 401:
        raise BoardError("session-expired", "Kilter sync unauthorized", "kilter")
    if not r.ok:
        raise BoardError("unexpected-response", f"sync failed ({r.status_code})", "kilter")

    tables: dict = {}
    try:
        for line in r.iter_lines(decode_unicode=True):
            if not line:
                continue
            try:
                msg = json.loads(line)
            except ValueError:
                continue
            if msg.get("data"):
                _apply_bucket_data(tables, msg["data"])
            if "checkpoint_complete" in msg:
                break
    finally:
        r.close()
    return tables


def _apply_bucket_data(tables: dict, sync_data: dict) -> None:
    for op in sync_data.get("data") or []:
        object_type = op.get("object_type")
        object_id = op.get("object_id")
        if not object_type or not object_id:
            continue
        table = tables.setdefault(object_type, {})
        if op.get("op") == "REMOVE":
            table.pop(object_id, None)
            continue
        raw = op.get("data")
        if isinstance(raw, str):
            try:
                table[object_id] = json.loads(raw)
            except ValueError:
                table[object_id] = {"_raw": raw}
        elif isinstance(raw, dict):
            table[object_id] = raw
        else:
            table[object_id] = {}
